Adding doc on asset objects (#21)
podium-lib/podium-lib.github.io: docs/api/assets.md

---
id: assets
title: Assets
---

When an asset is registered through the `.css()` or `.js()` methods in a podlet

> or layout an appropiate `AssetCSS` or `AssetJS` object is created.

Reviewer: ```suggestion
or layout an appropriate `AssetCSS` or `AssetJS` object is created.
```

        const incoming = request.app.podium;

        console.log(incoming.css)  // array of AssetCSS objects
        console.log(incoming.js)   // array of AssetJS objects

        [ ... ]
    },
});
```

<!--Fastify-->

```js
const podlet = new Podlet([ ... ]);

podlet.css({ value: '/assets/styles.css' });
podlet.js({ value: '/assets/scripts.js', type: 'esm' });

app.get(podlet.content(), async (request, reply) => {
    const incoming = reply.app.podium;

    console.log(incoming.css)  // array of AssetCSS objects
    console.log(incoming.js)   // array of AssetJS objects

    [ ... ]
});
```

<!--HTTP-->

```js
const podlet = new Podlet([ ... ]);

podlet.css({ value: '/assets/styles.css' });
podlet.js({ value: '/assets/scripts.js', type: 'esm' });

const server = http.createServer(async (req, res) => {
    let incoming = new HttpIncoming(req, res);
    incoming = await podlet.process(incoming);

    console.log(incoming.css)  // array of AssetCSS objects
    console.log(incoming.js)   // array of AssetJS objects

    [ ... ]
});

server.listen(7100);
```

<!--END_DOCUSAURUS_CODE_TABS-->

When a layout fetches a podlet and parses the podlets manifest, any assets on

> the `.css` and `.js` properties in the manifest is also parsed into appropiate

Reviewer: ```suggestion
the `.css` and `.js` properties in the manifest are also parsed into appropriate
```

> `AssetCSS` or `AssetJS` object.

Reviewer: ```suggestion
`AssetCSS` or `AssetJS` objects.
```

        console.log(response.css)  // array with the podlets AssetCSS objects
        console.log(response.js)   // array with the podlets AssetJS objects

        [ ... ]
    },
});

app.start();
```

<!--Fastify-->

```js
const podlet = layout.client.register({
    name: 'myPodlet',
    uri: 'http://localhost:7100/manifest.json',
});

app.get(layout.pathname(), async (request, reply) => {
    const incoming = reply.app.podium;

    const response = await podlet.fetch(incoming);

    console.log(response.css)  // array with the podlets AssetCSS objects
    console.log(response.js)   // array with the podlets AssetJS objects

    [ ... ]
});
```

<!--HTTP-->

```js
const podlet = layout.client.register({
    name: 'myPodlet',
    uri: 'http://localhost:7100/manifest.json',
});

const server = http.createServer(async (req, res) => {
    let incoming = new HttpIncoming(req, res);
    incoming = await layout.process(incoming);

    const response = await podlet.fetch(incoming);

    console.log(response.css)  // array with the podlets AssetCSS objects
    console.log(response.js)   // array with the podlets AssetJS objects

    [ ... ]
});
```

<!--END_DOCUSAURUS_CODE_TABS-->

> When a layout fetches one or more podlets its common that one would like to

Reviewer: ```suggestion
When a layout fetches one or more podlets it's common that one would like to
```

include the assets to these podlets in the full HTML document the layout is
composing.

By setting the [`Podlet Response`](layout.md#podlet-response) objects from the
podlets one fetch on to the `HttpIncoming.podlets` property in a layout one
appends the `AssetCSS` and `AssetJS` objects for the requested podlets onto the
`HttpIncoming.css` and `HttpIncoming.js` properties.

> This way; the layouts registered assets and the assets for all the requested

Reviewer: ```suggestion
This way the layouts registered assets and the assets for all the requested
```
The `;` should be used when two sentences can be both stand-alone, and separated by a comma.

| crossorigin | `string`  | &check; | &check; | `undefined`  | Correlates to the same attribute on a HTML `<link>` element |
| disabled    | `boolean` | &check; | &check; | `false`      | Correlates to the same attribute on a HTML `<link>` element |
| hreflang    | `string`  | &check; | &check; | `undefined`  | Correlates to the same attribute on a HTML `<link>` element |
| title       | `string`  | &check; | &check; | `undefined`  | Correlates to the same attribute on a HTML `<link>` element |
| media       | `string`  | &check; | &check; | `undefined`  | Correlates to the same attribute on a HTML `<link>` element |
| type        | `string`  | &check; | &check; | `text/css`   | Correlates to the same attribute on a HTML `<link>` element |
| rel         | `string`  | &check; | &check; | `stylesheet` | Correlates to the same attribute on a HTML `<link>` element |
| as          | `string`  | &check; | &check; | `undefined`  | Correlates to the same attribute on a HTML `<link>` element |

## Methods

An `AssetCSS` instance has the following methods:

### .toJSON()

> Returns JSON representation of the `AssetCSS` instance.

Reviewer: ```suggestion
Returns a JSON representation of the `AssetCSS` instance.
```

### .toHTML() {

Returns a HTML `link` element as a string of the `AssetCSS` instance.

## AssetJS

An `AssetJS` instance holds information about a Javascript related to a podlet

> or layout.

Reviewer 2: An `AssetJS` instance holds information about a podlet or layout's Javascript client side assets.

## Properties

An `AssetJS` instance has the following properties:

> | property    | type      | getter  | setter  | default      | details                                                     |

Reviewer 3: Looks like copy paste from CSS?

| ----------- | --------- | ------- | ------- | ------------ | ----------------------------------------------------------- |
| value       | `string`  | &check; |         | `''`         | Relative or absolute URL to the CSS asset                   |
| href        | `string`  | &check; |         | `''`         | Alias for the `value` property                              |

> | crossorigin | `string`  | &check; | &check; | `undefined`  | Correlates to the same attribute on a HTML `<link>` element |

Reviewer 2: an HTML

| disabled    | `boolean` | &check; | &check; | `false`      | Correlates to the same attribute on a HTML `<link>` element |
| hreflang    | `string`  | &check; | &check; | `undefined`  | Correlates to the same attribute on a HTML `<link>` element |
| title       | `string`  | &check; | &check; | `undefined`  | Correlates to the same attribute on a HTML `<link>` element |
| media       | `string`  | &check; | &check; | `undefined`  | Correlates to the same attribute on a HTML `<link>` element |
| type        | `string`  | &check; | &check; | `text/css`   | Correlates to the same attribute on a HTML `<link>` element |
| rel         | `string`  | &check; | &check; | `stylesheet` | Correlates to the same attribute on a HTML `<link>` element |
| as          | `string`  | &check; | &check; | `undefined`  | Correlates to the same attribute on a HTML `<link>` element |

## Methods

An `AssetJS` instance has the following methods:

### .toJSON()

> Returns JSON representation of the `AssetJS` instance.

Reviewer 2: a JSON

> Returns JSON representation of the `AssetJS` instance.

Reviewer: ```suggestion
Returns a JSON representation of the `AssetJS` instance.
```

### .toHTML() {

> Returns a HTML `script` element as a string of the `AssetJS` instance.

Reviewer 2: Returns an HTML `script` element as a string representation of the AssetJS instance.
